### PR TITLE
errhan: improve handling error code from user callbacks

### PR DESCRIPTION
**Pull request by dalcinl**
## Pull Request Description

Intended behavior:

* If a user callback returns an non-MPI error code, put it in the error
  string but do not return it back to the caller of the outer MPI routine.
  Otherwise, inspecting a non-MPI error code for error class or string
  is doomed to fail.

* If a user callback returns a predefined error class, still create and
  return error code of that class. This way, the error string will contain
  the stack trace, and the caller can still get the error class out of
  the returned error code.

* Dynamically defined error classes/codes pass through from callbacks to
  the caller of the outer MPI routine. Unfortunately, inspecting the error
  string will not provide the stack trace.

Notes:

* The fist commit is preparatory, consolidating bitmask operations, there is not change in behavior.
* The second commit is the juicy one, it implements the behavior described above.
* The third `fixup!` commit is just removal of an extra level of indentation after the previous changes, there is not change in behavior.
* The fourth commit add tests for the behavior implemented in the second.

Whether to keep these changes in separate commits or squash them is up to the MPICH maintainers.

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

File: src/mpi/errhan/errutil.c [hunk 1]
```diff
@@ -1061,18 +1060,23 @@ int MPIR_Err_create_code_valist(int lastcode, int fatal, const char fcname[],
...
                     specific_fmt = specific_err_msgs[specific_idx].long_name;
                 } else {
                     specific_fmt = specific_msg;
                 }
+                user_error_code = va_arg(Argp, int);
+                if (user_error_code & ERROR_DYN_MASK) {
+                    use_user_error_code = 1;
+                } else if (is_valid_error_class(user_error_code)) {
+                    error_class = MPIR_ERR_GET_CLASS(user_error_code);
+                }
```
Comment (hzhou) [inline src/mpi/errhan/errutil.c:1076]: So if user callback return a specific code that falls in the `else` branch, we just overwrite? What if the code has a specific meaning to the user?

Is the motivation from mpi4py always inspecting a return code using e.g. `MPI_Error_string` and you can't tell if it is from a user callback? I think the proper way is for mpi4py to wrap a user callback return with a specific (or dynamic) error code before passing thru MPI. That way, you ensure raw user code never gets to MPI and you can design your own mechanism of preserving user code.

Comment (dalcinl) [reply src/mpi/errhan/errutil.c:1076]: > So if user callback return a specific code that falls in the `else` branch, we just overwrite?

Yes. Although the exact value will end up in the error string.

>  What if the code has a specific meaning to the user?

Then the user should rather register and return a dynamic error code.

> Is the motivation from mpi4py always inspecting a return code using e.g. `MPI_Error_string` and you can't tell if it is from a user callback? 

You have to understand that I do not do things just because of silly and frivolous mpi4py motivations.

To your specific question, no, I do not by default call `MPI_Error_string`. But that could eventually happen, if an uncaught exception is printed in the the traceback.

Checking every MPI call for errors and eventually querying the error string  is not something exclusive of mpi4py. PETSc also checks the return code of MPI calls, and gets the error string to embed that information in their own error code handling.

> I think the proper way is for mpi4py to wrap a user callback return with a specific (or dynamic) error code before passing thru MPI. That way, you ensure raw user code never gets to MPI and you can design your own mechanism of preserving user code.

That's what mpi4py already does, and has done that for years! Python user callbacks are executed in a `try/except` block, and any exception is translated to returning error code. Actually, I'm not even that exactly because you refused my interpretation of delete callbacks as destructors for attribute values.

Again, this is not about Python and mpi4py!!! The root of the problem is that an outer MPI routine returning a non-MPI error code not know to MPI is a bomb in user hands waiting to explode: there is no standard MPI way to tell whether the error code is a valid MPI error code, and calling `MPI_Error_string` on such error code is expected to trigger to an additional failure (that is, asking the error string of an unknown error code), thus swallowing the original point of failure.

Long story short: In my opinion, allowing MPI routines to return unknown MPI error codes is bad, simply because it does not compose well across third-party library boundaries. If libraryA creates a generalized request that ends up in the hands of libraryB calling `MPI_Wait` on it, and a grequest callback returns a non MPI error code, then the return value is a) meaningless to libraryB, and b) useless because an attempt to `MPI_Error_string` would either fail or return garbage. Therefore, non-MPI errors should not be let pass through. This is the heart of the problem I'm trying to address. Can we focus on that and forget about Python and mpi4py?

Comment (hzhou) [reply src/mpi/errhan/errutil.c:1076]: > The root of the problem is that an outer MPI routine returning a non-MPI error code not know to MPI is a bomb in user hands waiting to explode

The "bomb" originates from the user and returns back to the same user to consume. It is just passing thru MPI. It may be a bomb, or may be part of user's meaningful design. So my philosophy is we handle it with care and return back to user and trust user know what they are doing. There are obviously the case when users don't know what they are doing, but since we don't know that, too eager to help may end up messing up "expert" users.

Comment (dalcinl) [reply src/mpi/errhan/errutil.c:1076]: > The "bomb" originates from the user and returns back to the same user to consume.

Not necessarily. I already mentioned libraryA returning a non-MPI error code, and libraryB having to deal with that unknown error code. What's your proposal for such scenario?

Comment (hzhou) [reply src/mpi/errhan/errutil.c:1076]: Where does this non-MPI error code originate from, libraryA or libraryB?

Comment (dalcinl) [reply src/mpi/errhan/errutil.c:1076]: I already gave an example involving generalized requests. LibraryA returns a generalized request, let say the query function (implemented in LibraryA) may eventually returns a non-MPI error code. The generalized request is passed to LibraryB, and eventually LibrayyB will wait for the request. The waiting triggers the invocation of the query callback (in LbraryA land), that then returns a non MPI error code that gets propagated to the Wait call in (in LibraryB land). 

How is LibaryB supposed to deal with the non-MPI return code generated in LibraryA? Of course, you may argue that LibraryA should document things (or fix their code to do it the MPI way), and LibraryB should be prepared for the LibraryA behavior. But sometimes that's very hard to achieve in practice, simply because modern software is composed of many layers involving many independent libraries.

Comment (hzhou) [reply src/mpi/errhan/errutil.c:1076]: In this case the non-MPI error originates in LibraryA (from the query function) and is passing through libraryB. LibraryB should manage to pass the error back to libraryA since libraryA may depend on that special non-MPI error. How libraryB do that maybe complicated but it depends on the interface between libraryA and libraryB. You may argue that a `MPI_Wait` call is innocent. But since libraryB is waiting on a generalized request from libraryA, there must be an API design for that.

**Comment by dalcinl** [comment]
@hzhou I would like to put things crystal clear... 

Are you advocating for an MPI error handler model where some outer `MPI_Xxx_yyy()`` routine involving the invocation a user callback (e.g. `MPI_Wait()` or `MPI_Comm_dup()`) may fail and return an error code, but that error code MAY NOT be a proper MPI error code, moreover there is not way to check whether that error code is valid, and if you query `MPI_Error_string` or `MPI_Error_class` you may get a new error in turn saying that the error code is invalid and unknown to MPI ?

Am I accurately characterizing the position you are taking to reject what I've done in this PR?

**Comment by dalcinl** [comment]
> For example, a generalized request created by libraryA is owned by libraryA

If LibraryB is the one that what to wait on it, then ratherI consider than then libraryA transferred ownership of the request to libraryB. 

Sometimes errors may happen even if ownership is not transferred but borrowed.

For example, supposed that PETSc attaches some attribute to a communicator, and there is a copy_fn that may fail with a non-MPI error code for whatever reason. Then I get the MPI_Comm out of a PETSc object, and next I do `MPI_Comm_dup` because I really need a new comm context. But then copy_fn run, it fails, and I get an alien error code unknown to MPI. What ownership are we talking about?


> Thus, if you ignore the ownership discussion, then you are not accurately characterizing my position.

No matter what how you decorate your exposition, you are advocating for a model where outer MPI calls return error codes that are unknown to MPI and on which you cannot use the most basic MPI error reporting machinery based on `MPI_Error_class` and `MPI_Error_string`, because if you do, you get (obviously) a new and unrelated error about the unknown thing.

At this point, and once again, there is no point on keep the discussion going. I consider MPI returning unknown error codes unacceptable, no matter the circumstances. You do not share that view. You never ask for third-party opinions. You win.

**Comment by hzhou** [comment]
> If LibraryB is the one that what to wait on it, then ratherI consider than then libraryA transferred ownership of the request to libraryB.
> 
> Sometimes errors may happen even if ownership is not transferred but borrowed.

Borrowing something has the same side effects as transferring ownership. I am not saying it is trivial, but when libraryB borrows an object from libraryA without fully prepared to handle all side effects for that object, whose fault is it? I am also not saying this is good. Ideally we want as little side effects as possible. But since MPI allows user callbacks and allows user callbacks to return non-zero code, all layers that may get that user error code need have measures to deal with it. In principle, that user-originated code need somehow return back to the user to interpret since it may be part of user's deliberate design.

> 
> For example, supposed that PETSc attaches some attribute to a communicator, and there is a copy_fn that may fail with a non-MPI error code for whatever reason. Then I get the MPI_Comm out of a PETSc object, and next I do `MPI_Comm_dup` because I really need a new comm context. But then copy_fn run, it fails, and I get an alien error code unknown to MPI. What ownership are we talking about?

It is using an `MPI_Comm` owned by PETSc and the user need understand the side effects of it. Just like when you borrow a car for whatever reason (even for dipping)  you are responsible in understanding its quirks and be prepared for it.

An informed user should dup the comm before passing to PETSc. Or PETSc could require its users do not attach attributes that potential can fail a callback before init with the comm.

> 
> > Thus, if you ignore the ownership discussion, then you are not accurately characterizing my position.
> 
> No matter what how you decorate your exposition, you are advocating for a model where outer MPI calls return error codes that are unknown to MPI and on which you cannot use the most basic MPI error reporting machinery based on `MPI_Error_class` and `MPI_Error_string`, because if you do, you get (obviously) a new and unrelated error about the unknown thing.
> 
> At this point, and once again, there is no point on keep the discussion going. I consider MPI returning unknown error codes unacceptable, no matter the circumstances. You do not share that view. 

The culprit is MPI allow user callback to return non-zero and arbitrary error code and there is no clear spec on the restrictions. Returning error code that originates from user back to user is different from returning an error code that user does not understand and cannot query. In practice, the use case that user callbacks returning error code are rather rare, and I tend to believe that when user do that, they must be doing it for a very specific reason and we'd better not to break that reason even though we have no understanding of it.

Now flipping the table, have you tried consider my view? I ask this because you keep trying to rewrite my view into something that does not truly represent mine.

> You never ask for third-party opinions. 

Wait a minute. I have been open for third-party opinions (or any party's opinion including yours) the whole time. I did not insist I was right. I was merely stating my opinion/philosophy. Which part give you the end-of-discussion impression of mine?

If you mean to ask for third-party opinions as to appeal to authority, you know that is a logic fallacy, right?

> You win.
;) 
The discussion is never about win/lose. I never feel good if you just surrender without truly being convinced. It is perfectly okay to adjourn or abandon the discussion. On the other hand, if you are aiming to win, that is tough. How can you win if I am not in the game? But be assured that I do get convinced and often feel happy when that happens.

**Comment by hzhou** [comment]
@dalcinl I am closing this PR for now. The conversation is still open so feel free to continue or we can discuss offline. Please feel free to reopen the PR if you still wish for the change.
